### PR TITLE
feat: introduction of consolidation data in oracle report (BS-3863)

### PR DESCRIPTION
**Pull request by iamsahu**
## Description

This pull request introduces strict monotonicity checks and buffer management for the `totalConsolidationsAmountReported` field in the `OracleManagerV1` contract. It ensures that this value cannot decrease across reports, enforces that any increase does not exceed the available consolidation buffer, and updates the buffer accordingly. The changes are thoroughly tested with new and enhanced invariant and coverage tests.

**Core logic and validation enhancements:**

* Enforced that `totalConsolidationsAmountReported` cannot decrease in new reports and added a revert with a custom error if it does (`InvalidTotalConsolidationsAmountReportedDecrease`). If it increases, the increase cannot exceed the current consolidation buffer; otherwise, a revert with `InvalidTotalConsolidationsAmountReportedIncrease` occurs. When valid, the buffer is reduced by the increase. (`[[1]](diffhunk://#diff-ad4a762ff4e4eefd2b88f7a66eea1e8deaff75754db1856d9a855f9c68fa7b7aR329-R347)`, `[[2]](diffhunk://#diff-41d9715972834af03b494955cfa0b5440436a746d86707386bc46f164a62b0d6R100-R113)`)
* Updated storage and struct definitions to include `totalConsolidationsAmountReported` in both the in-memory and persistent report structures. (`[[1]](diffhunk://#diff-41d9715972834af03b494955cfa0b5440436a746d86707386bc46f164a62b0d6R154-R156)`, `[[2]](diffhunk://#diff-41d9715972834af03b494955cfa0b5440436a746d86707386bc46f164a62b0d6R192)`, `[[3]](diffhunk://#diff-ad4a762ff4e4eefd2b88f7a66eea1e8deaff75754db1856d9a855f9c68fa7b7aR383)`)

**Testing and invariants:**

* Added a new invariant (`invariant_I21_consolidationsAmountNonDecreasing`) to the accounting invariants test, ensuring `totalConsolidationsAmountReported` never decreases across reports. (`[[1]](diffhunk://#diff-66960b1856a37fa2e419010faf50c5ef224e56d8f7d255ddbb2634c632b0c8fcR20)`, `[[2]](diffhunk://#diff-66960b1856a37fa2e419010faf50c5ef224e56d8f7d255ddbb2634c632b0c8fcL90-R96)`, `[[3]](diffhunk://#diff-66960b1856a37fa2e419010faf50c5ef224e56d8f7d255ddbb2634c632b0c8fcR287-R302)`)
* Expanded the test harness and coverage tests to:
  - Expose and test buffer updates and monotonicity for consolidations.
  - Check all revert and update paths for consolidation amount changes, including fuzzing for edge cases. (`[[1]](diffhunk://#diff-390f6f7daefb40a4a01ac34eea5e6cc51027379a80d4c46ceff0f9bc7148e387R14-R26)`, `[[2]](diffhunk://#diff-390f6f7daefb40a4a01ac34eea5e6cc51027379a80d4c46ceff0f9bc7148e387R36-R39)`, `[[3]](diffhunk://#diff-390f6f7daefb40a4a01ac34eea5e6cc51027379a80d4c46ceff0f9bc7148e387R126-R129)`, `[[4]](diffhunk://#diff-390f6f7daefb40a4a01ac34eea5e6cc51027379a80d4c46ceff0f9bc7148e387R535)`, `[[5]](diffhunk://#diff-390f6f7daefb40a4a01ac34eea5e6cc51027379a80d4c46ceff0f9bc7148e387L632-R816)`)

**Internal consistency and storage:**

* Ensured buffer changes are observable in storage and included in the asset balance computation, mirroring the logic in `RiverV1`. (`[contracts/test/components/OracleManager.1.t.solR126-R129](diffhunk://#diff-390f6f7daefb40a4a01ac34eea5e6cc51027379a80d4c46ceff0f9bc7148e387R126-R129)`)

These changes collectively harden the protocol against invalid consolidation reporting, maintain accurate accounting, and provide robust test coverage for the new logic.

## Notice

- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/liquid-collective/liquid-collective-protocol/pulls) for the same update/change?
- [ ] Have you assigned this PR to yourself?
- [ ] Have you added at least 1 reviewer?
- [ ] Have you updated the official documentation?
- [ ] Have you added sufficient documentation in your code?
- [ ] Have you added relevant tests to the official test suite?

## Pull Request Type

- [ ] 💫 New Feature (Breaking Change)
- [ ] 💫 New Feature (Non-breaking Change)
- [ ] 🛠️ Bug fix (Non-breaking Change: Fixes an issue)
- [ ] 🕹️ Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)

## Breaking changes (if applicable)

breaks the format of the old oracle report

## Testing

- [ ] Have you tested this code with the official test suite?
- [ ] Have you tested this code manually?

### CODE REVIEW

File: contracts/src/components/OracleManager.1.sol [hunk 1]
```diff
@@ -258,6 +258,8 @@ abstract contract OracleManagerV1 is IOracleManagerV1 {
         uint256 skimmedAmountIncrease;
         uint256 inFlightDepositedETH;
         uint256 totalDepositedActivatedETHIncrease;
+        uint256 oldConsolidationBuffer;
```
Comment (juliaaschmidt) [inline contracts/src/components/OracleManager.1.sol]:
```suggestion
        uint256 lastConsolidationBuffer;
```
more consistent with the other fields

File: contracts/src/components/OracleManager.1.sol [hunk 2]
```diff
@@ -326,6 +328,33 @@ abstract contract OracleManagerV1 is IOracleManagerV1 {
...
+
+            if (
+                _report.totalExternalConsolidationsAmountReported
+                    > lastStoredReport.totalExternalConsolidationsAmountReported
+            ) {
+                // the total consolidation amount reported has increased so we need to reduce the buffer
+                uint256 increaseInConsolidation = _report.totalExternalConsolidationsAmountReported
+                    - lastStoredReport.totalExternalConsolidationsAmountReported;
+                vars.oldConsolidationBuffer = ConsolidationBuffer.get();
+                if (increaseInConsolidation > vars.oldConsolidationBuffer) {
```
Comment (juliaaschmidt) [inline contracts/src/components/OracleManager.1.sol]: How accurately will these values be reported, especially considering the difference between effective and actual balance? I’m wondering whether this risks to consistently revert here because of the issue described in the Notion doc: https://app.notion.com/p/glxydg/External-Consolidations-Potential-Consolidated-Amount-Discrepancy-3428d6bc7ac380afa141e419ff3b7fef?source=copy_link

File: contracts/src/interfaces/components/IOracleManager.1.sol [hunk 3]
```diff
@@ -137,6 +151,10 @@ interface IOracleManagerV1 {
         // this value cannot decrease over reports
         // this value includes only the ETH that was deposited on the Execution Layer Deposit contract
         uint256 totalDepositedActivatedETH;
+        // the sum of all external consolidation funds that were reported
+        // this will only include the consolidations that happend due to external validators merging into LC
```
[the remark this answers is not in the document]
Comment (juliaaschmidt) [reply contracts/src/interfaces/components/IOracleManager.1.sol]: please fix

File: contracts/src/AttestationVerifier.1.sol [hunk 4]
```diff
@@ -147,6 +147,18 @@ contract AttestationVerifierV1 is Initializable, IAttestationVerifierV1 {
             revert QuorumExceedsMaxSignatures(_consolidationQuorum, MAX_SIGNATURES);
         }
 
+        // ---- Validate consolidation-side parameters ----
+        if (
```
Comment (juliaaschmidt) [inline contracts/src/AttestationVerifier.1.sol]: this is here by mistake. This is code duplication from line 138-148. Probably due to merge conflicts resolution sorry.

**Review by juliaaschmidt — CHANGES_REQUESTED** [review]

File: contracts/src/AttestationVerifier.1.sol (hunk 4 above)
Comment (iamsahu) [reply contracts/src/AttestationVerifier.1.sol]: Fixed

File: contracts/src/components/OracleManager.1.sol (hunk 2 above)
Comment (mischat) [reply contracts/src/components/OracleManager.1.sol]: **Confirming this concern — I'd rate it HIGH (liveness). The revert here can brick *all* oracle reporting.**

**Root cause: two independent off-chain numbers must agree on-chain, but the CL guarantees they can diverge.**

- **Mint side** increments the buffer by the *committee-attested* `consolidation.totalAmount` — the source validator's **effective balance (EB) at CL-processing time** (`River.1.sol` L216-217):
  ```solidity
  _setConsolidationBuffer(oldConsolidationBuffer, oldConsolidationBuffer + consolidation.totalAmount);
  _mintShares(recipient, consolidation.totalAmount);
  ```
- **Oracle side** (here) decrements by the *actual ETH that landed*, and reverts if `increase > buffer`.

Per the CL spec / the Notion doc linked above, the amount actually moved at `process_pending_consolidations` is **`min(actual_balance, effective_balance)` at execution time**, not the EB snapshotted at request time. While the source sits in the exit queue it keeps earning: a validator just below an EB hysteresis boundary can step EB **up ~1 ETH** and consolidate **more than was minted/buffered**.

So `increaseInConsolidation > lastConsolidationBuffer` → `revert InvalidTotalConsolidationsAmountReportedIncrease` → **the entire `setConsensusLayerData` call reverts**. Since reporting drives rewards/exits/redemptions/rebalancing, one consolidation that out-earns its snapshot **halts the protocol** until the buffer is manually reconciled — no in-contract recovery.

This also implies an **undocumented contract⇄oracle invariant** ("oracle must cap its reported delta at the buffered amount and surface the surplus as ordinary rewards") that is stated nowhere.

**Recommended fix — clamp instead of revert:**
```solidity
uint256 bufferReduction = LibUint256.min(increaseInConsolidation, vars.lastConsolidationBuffer);
vars.totalExternalConsolidationsAmountReportedIncrease = bufferReduction;
```
- reward case (`actual > minted`): buffer → 0; surplus booked as rewards, already gated by the `TotalValidatorBalanceIncreaseOutOfBound` APR ceiling → bounded, no brick.
- penalty case (`actual < minted`): unchanged — residual buffer drained by the consolidation-coverage-fund pull at L483-495.
- a buggy/over-reporting oracle is then capped by the APR bound rather than able to halt reporting — strictly safer than the hard revert.

**Minimum if the revert is kept:** document the `reported ≤ buffer` invariant on the function + struct, and replace the "exceeds buffer reverts" test with a divergent-amount test (rewards-during-exit-queue at an EB boundary) asserting the intended outcome.

File: contracts/test/accounting/BeaconChainSimulator.sol [hunk 5]
```diff
@@ -249,11 +254,13 @@ abstract contract BeaconChainSimulator is AccountingHarnessBase {
...
         }
 
-        report.validatorsBalance = validatorsBalance;
+        // The consolidated principal has landed on the CL, so it is part of the reported validatorsBalance.
+        report.validatorsBalance = validatorsBalance + _simConsolidatedBalance;
         report.validatorsSkimmedBalance = _simCumulativeSkimmed;
         report.validatorsExitedBalance = _simCumulativeExited;
         report.validatorsExitingBalance = validatorsExiting;
         report.totalDepositedActivatedETH = _simTotalDepositedActivatedETH;
+        report.totalExternalConsolidationsAmountReported = _simConsolidatedBalance;
```
Comment (mischat) [inline contracts/test/accounting/BeaconChainSimulator.sol:263]: **H-1 test gap (see thread on `OracleManager.1.sol` L349).** Both legs here are driven by the same `_simConsolidatedBalance`:

```solidity
report.validatorsBalance = validatorsBalance + _simConsolidatedBalance;
report.totalExternalConsolidationsAmountReported = _simConsolidatedBalance;
```

so the simulator always produces `reported == buffered` and can **never** exercise the `reported > buffered` path that triggers the revert in `setConsensusLayerData`. The HIGH liveness case (source validator earns rewards in the exit queue → consolidates ~1 ETH+ above the minted/buffered amount) is therefore untested. Please add a divergent-amount scenario at an EB hysteresis boundary.

File: contracts/test/accounting/scenarios/ConsolidationCoverage.t.sol [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        assertEq(uint256(vm.load(address(river), CONSOLIDATION_BUFFER_SLOT)), 0, "buffer drawn down by delta");
+        // Core invariant: the consolidated principal is exactly offset, so the total underlying is UNCHANGED
+        // (neither a drop nor an increase) and no fee shares are minted on the principal.
+        assertEq(river.totalUnderlyingSupply(), underlyingBefore, "total underlying unchanged (netted out)");
+        assertEq(river.totalSupply(), sharesBefore, "no fee shares minted on consolidated principal");
+    }
+
+    /// @notice A reported consolidation increase larger than the available buffer reverts through the
+    ///         real River -> Oracle flow with InvalidTotalConsolidationsAmountReportedIncrease.
+    function testConsolidationIncreaseExceedingBufferReverts() public {
```
Comment (mischat) [inline contracts/test/accounting/scenarios/ConsolidationCoverage.t.sol]: Tying into the H-1 thread on `OracleManager.1.sol` L349: this test codifies the **brick** (`reported > buffer` reverts) as intended behaviour. If we adopt the clamp fix (decrement by `min(increase, buffer)`, surplus → rewards), this should become a *netting* test instead: `validatorsBalance += actual` with `actual > buffered` at an EB boundary, asserting the surplus is booked as APR-bounded rewards and reporting does **not** revert.

**Review by juliaaschmidt — DISMISSED** [review]
LGTM

File: contracts/src/components/OracleManager.1.sol (hunk 2 above)
Comment (mischat) [reply contracts/src/components/OracleManager.1.sol]: ✅ **Confirmed resolved by `75c5c41a`** — this is exactly the clamp fix proposed above. The hard revert is replaced with `bufferReduction = min(increase, buffer)`:

```solidity
if (increaseInConsolidation > vars.lastConsolidationBuffer) {
    // buffer fully covered, extra goes to rewards (already in the validatorsBalance increase)
    vars.totalExternalConsolidationsAmountReportedIncrease = vars.lastConsolidationBuffer;
} else {
    vars.totalExternalConsolidationsAmountReportedIncrease = increaseInConsolidation;
}
```

**Why it's correct** (`post − pre = validatorsBalanceΔ − bufferReduction`):
- `increase ≤ buffer` (normal): `increase − increase = 0` → not booked as rewards.
- `increase > buffer` (the reward case that used to brick): buffer → 0, surplus `increase − buffer` surfaces in `validatorsBalance` as ordinary rewards, gated by the existing `TotalValidatorBalanceIncreaseOutOfBound` APR ceiling → no protocol-wide brick.
- penalty case unchanged — residual buffer still drained by the consolidation-coverage-fund pull.

The stored monotonic `totalExternalConsolidationsAmountReported` still records the full reported value (future deltas keep tracking actual amounts) while only the clamped amount leaves the buffer — self-correcting. The only remaining revert is the general APR bound, which is the intended safety mechanism.

**Tests** also addressed the gap: both brick-codifying tests removed, and `testFuzzSetConsensusLayerDataConsolidations` now exercises the divergent `delta > buffer` path — caps reduction at `min(delta, buffer)`, asserts surplus → rewards, and asserts `TotalValidatorBalanceIncreaseOutOfBound` fires when the surplus exceeds `maxIncrease`. Locally: `forge build` clean, 9/9 consolidation tests pass (1500 fuzz runs).

H-1 resolved 👍. Only remaining item from my side is the **doc-only M-2** (worth a NatSpec note: the netting assumes the consolidation field is bumped in the *same* report whose `validatorsBalance` reflects the landed principal — now more relevant since any surplus routes through the APR bound). Not a blocker.

File: contracts/test/accounting/BeaconChainSimulator.sol (hunk 5 above)
Comment (mischat) [reply contracts/test/accounting/BeaconChainSimulator.sol:263]: Tracked as #574. The clamp fix (this thread's root concern on `OracleManager.1.sol:349`) and the brick→netting test conversion (`ConsolidationCoverage.t.sol:142`) are resolved; #574 captures the remaining gap raised here — the simulator hard-couples `validatorsBalance` and `totalExternalConsolidationsAmountReported` (L258/L263), so the realistic `reported > buffered` divergence at an EB hysteresis boundary still isn't exercised end-to-end (only unit-fuzzed). Leaving this thread for that follow-up.

Comment (mischat) [reply contracts/test/accounting/BeaconChainSimulator.sol:263]: happy to push this to the issue and resolve it in the future.

File: contracts/test/accounting/scenarios/ConsolidationCoverage.t.sol (hunk 6 above)
Comment (mischat) [reply contracts/test/accounting/scenarios/ConsolidationCoverage.t.sol]: this is fixed.

File: contracts/src/components/OracleManager.1.sol (hunk 2 above)
Comment (mischat) [reply contracts/src/components/OracleManager.1.sol]: Tracked as #575 (doc-only). It captures the NatSpec note proposed here — documenting the same-report invariant (`totalExternalConsolidationsAmountReported` must be bumped in the same report whose `validatorsBalance` reflects the landed principal) on the struct field (`IOracleManager.1.sol:157`) and the clamp block (`OracleManager.1.sol:345`), including the post-clamp consequence that any surplus now routes through the APR bound. Not a blocker for this PR.

Comment (mischat) [reply contracts/src/components/OracleManager.1.sol]: i think we can resolve this now.

**Review by mischat — DISMISSED** [review]
I have opened two issues for the future, they are both minor, docs related etc ...

**Review by mischat — APPROVED** [review]
:shipit:

**Review by juliaaschmidt — APPROVED** [review]
